canvas/gradient.go

```diff
@@ -0,0 +1,190 @@
+package canvas
+
+import (
+	"image"
+	"image/color"
+	"image/draw"
+	"math"
+
+	"fyne.io/fyne"
+)
+
+// Gradient describes a gradient between two colors
+type Gradient struct {
+	baseObject
+
+	Generator func(w, h int) image.Image
+
+	Translucency float64 // Set a translucency value > 0.0 to fade the raster
+
+	img draw.Image // internal cache for pixel generator
+}
+
+// Alpha is a convenience function that returns the alpha value for a raster
+// based on its Translucency value. The result is 1.0 - Translucency.
+func (g *Gradient) Alpha() float64 {
+	return 1.0 - g.Translucency
+}
+
+// gradient generates a pixel using the defined gradient
+// function by using the w, h, and current x,y  of this pixel
+func gradient(gradientFnc func(x, y, ox, oy, w, h int) float64, start color.Color, end color.Color, w, h, ox, oy, x, y int) *color.RGBA64 {
+	d := gradientFnc(x, y, ox, oy, w, h)
+	// fetch RGBA values
+	aR, aG, aB, aA := start.RGBA()
+	bR, bG, bB, bA := end.RGBA()
+
+	// Get difference
+	dR := (float64(bR) - float64(aR))
+	dG := (float64(bG) - float64(aG))
+	dB := (float64(bB) - float64(aB))
+	dA := (float64(bA) - float64(aA))
+
+	// Apply graduation
+	pixel := &color.RGBA64{
+		R: uint16(float64(aR) + d*dR),
+		B: uint16(float64(aB) + d*dB),
+		G: uint16(float64(aG) + d*dG),
+		A: uint16(float64(aA) + d*dA),
+	}
+
+	// Prevent rendering dead space - transparent corners
+	if d > 1 {
+		pixel.A = uint16(0)
+	}
+
+	return pixel
+}
+
+type pixelGradient struct {
+	g   *Gradient
+	img draw.Image
+}
+
+// NewRectangleGradient returns a new Image instance that dynamically
+// renders a gradient based off of options
+// Accepts
+func NewRectangleGradient(optFnc ...GradientOption) *Gradient {
+	options := &GradientOptions{}
```

Comment:
As noted below I think this would be simpler without the options pattern.
None of the other primitives use this technique as their functionality is fairly well defined with few optionals.